Compiler bug: "too many arguments in call" on MacOS

### Describe the bug

Errors compiling C code emitted by V: `error: too many arguments in call`. The error only manifests in MacOS (Linux and Windows pass).

### Expected Behavior

No C compile errors.

### Current Behavior

I don't have access to a macOS so I can only report the Github Actions error.
Both Linux and MacOS run the same Github actions. Here are the Github actions test run logs:

https://github.com/srackham/v-rimu/actions/runs/4603691234/jobs/8226161609

Here's an error from the above logs:

```
FAIL  [ 3/17]   731.845 ms /Users/runner/work/v-rimu/v-rimu/v-rimu/delimitedblocks/delimitedblocks_test.v
==================
/tmp/v_501/tsession_117b51600_168750688/delimitedblocks_test.12765202679951327310.tmp.c:3854:55: error: too many arguments in call to 'anon_fn_array_string__bool_str' [-Werror]
        string _t3 = anon_fn_array_string__bool_str(it.verify);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          ^
/tmp/v_501/tsession_117b51600_168750688/delimitedblocks_test.12765202679951327310.tmp.c:3855:95: error: too many arguments in call to 'anon_fn_array_string_delimitedblocks__definition__string_str' [-Werror]
        string _t4 = anon_fn_array_string_delimitedblocks__definition__string_str(it.delimiter_filter);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                    ^
/tmp/v_501/tsession_117b51600_168750688/delimitedblocks_test.12765202679951327310.tmp.c:3856:91: error: too many arguments in call to 'anon_fn_string_array_string_expansion__options__string_str' [-Werror]
        string _t5 = anon_fn_string_array_string_expansion__options__string_str(it.content_filter);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                  ^
3 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

The errors center around compiled C functions such as `anon_fn_array_string__bool_str`, (compiled on Linux):

```
static string anon_fn_array_string__bool_str(); // auto
  :
static string anon_fn_array_string__bool_str() { return _SLIT("fn ([]string) bool");}
  :
	string _t3 = anon_fn_array_string__bool_str(it.verify);
```

### Reproduction Steps

I don't have access to macOS so haven't been able to debug this problem and narrow the issue down, so unfortunately there's not a lot to go on.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 cd629cf

### Environment details (OS name and version, etc.)

macOS 12.6.4 21G526 running on Github

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -189,6 +189,8 @@ fn (mut g Gen) gen_str_for_option(typ ast.Type, styp string, str_fn_name string)
 		g.auto_str_funcs.writeln('\t\tres = ${str_intp_sq(tmp_res)};')
 	} else if should_use_indent_func(sym.kind) && !sym_has_str_method {
 		g.auto_str_funcs.writeln('\t\tres = indent_${parent_str_fn_name}(*(${sym.cname}*)it.data, indent_count);')
+	} else if sym.kind == .function {
+		g.auto_str_funcs.writeln('\t\tres = ${parent_str_fn_name}();')
 	} else {
 		g.auto_str_funcs.writeln('\t\tres = ${parent_str_fn_name}(*(${sym.cname}*)it.data);')
 	}

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -162,7 +162,8 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 				g.write('*')
 			}
 			g.expr_with_cast(expr, typ, typ)
-		} else {
+		} else if typ.has_flag(.option) {
+			// only Option fn receive argument
 			g.expr_with_cast(expr, typ, typ)
 		}
 		g.write(')')

--- a/vlib/v/tests/print_fn_test.v
+++ b/vlib/v/tests/print_fn_test.v
@@ -1,0 +1,15 @@
+struct Foo {
+mut:
+	f ?fn (int)
+}
+
+fn test_print_fn() {
+	a := fn (a string) int {
+		return 1
+	}
+	println(a)
+
+	foo := Foo{}
+	println(foo.f)
+	assert foo.f == none
+}

